File: Computer Science/Algorithms/Branch and Bound/Topological Sort/code.py
# G[i][j] indicates whether the path from the i-th node to the j-th node exists or not.
# NOTE: The graph must be Directed-Acyclic
from collections import deque

G = [
    [0, 0, 0, 0, 0, 0],
    [0, 0, 1, 0, 0, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 0, 0],
    [1, 1, 0, 0, 0, 0],
]


inDegrees = [0 for _ in range(len(G))]
Q = deque()
iter = 0

print('Calculating in-degrees for each Node...')

for currNode in range(0, len(G)):
    for currNodeNeighbor in range(0, len(G)):
        if G[currNode][currNodeNeighbor]:
            inDegrees[currNodeNeighbor] += 1
            # print(f"{currNode} -> {currNodeNeighbor}")

print('Done.In - Degrees are: [ ${String(inDegrees)}]')
print('Initializing queue with all the sources (nodes with no incoming edges)')

for node, inDegree in enumerate(inDegrees):
    if not inDegree:
        print(f"{node} is a source")
        Q.append(node)

print(f"Done. Initial State of Queue: {Q}")

File: computer-science/algorithms/Branch and Bound/Topological Sort/code.js
// G[i][j] indicates whether the path from the i-th node to the j-th node exists or not. NOTE: The graph must be Directed-Acyclic
const G = [
  [0, 0, 0, 0, 0, 0],
  [0, 0, 1, 0, 0, 0],
  [0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0],
  [1, 0, 0, 1, 0, 0],
  [1, 1, 0, 0, 0, 0],
];

const inDegrees = Array(...Array(G.length)).map(Number.prototype.valueOf, 0); // create an Array of G.length number of 0s
const Q = [];
let iter = 0;
let i;

console.log('Calculating in-degrees for each Node...');

for (let currNode = 0; currNode < G.length; currNode++) {
  for (let currNodeNeighbor = 0; currNodeNeighbor < G.length; currNodeNeighbor++) {
    if (G[currNode][currNodeNeighbor]) {
      
      inDegrees[currNodeNeighbor]++;
      
    }
  }
}

console.log(`Done. In-Degrees are: [ ${String(inDegrees)} ]`);
console.log('');

console.log('Initializing queue with all the sources (nodes with no incoming edges)');

inDegrees.map((indegrees, node) => {
  
  if (!indegrees) {
    
    console.log(`${node} is a source`);
    
    Q.push(node);
  }
  
});

console.log(`Done. Initial State of Queue: [ ${String(Q)} ]`);
console.log('');

// begin topological sort (kahn)
while (Q.length > 0) {
  
  console.log(`Iteration #${iter}. Queue state: [ ${String(Q)} ]`);
  
  const currNode = Q.shift();

  for (i = 0; i < G.length; i++) {
    if (G[currNode][i]) {
      
      inDegrees[i]--;

      if (!inDegrees[i]) {
        
        console.log(`${i}'s in-degree is now 0. Enqueuing ${i}`);
        
        Q.push(i);
      }
    }
  }

  console.log(`In-degrees are: [${String(inDegrees)} ]`);
  console.log('-------------------------------------------------------------------');

  iter++;
}
